cmake/modules/extensions.cmake

```diff
@@ -5275,3 +5292,91 @@ function(add_llext_target target_name)
     DEPENDS ${llext_pkg_output}
   )
 endfunction()
+
+# Usage:
+#   add_llext_command(
+#     TARGET <target_name>
+#     PRE_BUILD | POST_BUILD | POST_PKG
+#     COMMAND <command> [...]
+#   )
+#
+# Add a custom command to an llext target that will be executed during
+# the build.
+#
+# The differrent build steps are:
+# - PRE_BUILD:  Before the llext code is linked.
+# - POST_BUILD: After the llext code is built, but before packaging
+#               it in an .llext file.
+# - POST_PKG:   After the .llext file has been created.
+#
+# Anything else after COMMAND will be passed to add_custom_command() as-is
+# (including multiple commands and other options).
+function(add_llext_command)
+  set(options     PRE_BUILD POST_BUILD POST_PKG)
+  set(single_args TARGET)
+  # COMMAND and other options are passed to add_custom_command() as-is
+
+  cmake_parse_arguments(PARSE_ARGV 0 LLEXT "${options}" "${single_args}" "${multi_args}")
+
+  # Get target name
+  if(NOT LLEXT_TARGET)
+    message(FATAL_ERROR "add_llext_command: TARGET argument must be provided")
+  endif()
+  set(target_name ${LLEXT_TARGET})
+
+  # Check the target exists and refers to an llext target
+  set(llext_lib_target  ${target_name}_llext_lib)
+  set(llext_proc_target ${target_name}_llext_proc)
+  if(NOT TARGET ${llext_lib_target})
+    message(FATAL_ERROR "add_llext_command: not an llext target: ${target_name}")
+  endif()
+
+  # ARM uses an object file representation so there is no link step. Both PRE_BUILD
+  # and POST_BUILD mean exactly the same thing in this case, and POST_BUILD is more
+  # appropriate.
+  if(CONFIG_ARM AND LLEXT_PRE_BUILD)
+    message(WARNING
+	    "add_llext_command: PRE_BUILD on this arch is the same as POST_BUILD")
+    set(LLEXT_PRE_BUILD FALSE)
+    set(LLEXT_POST_BUILD TRUE)
+  endif()
+
+  # Determine the build step and the target to attach the command to
+  # based on the provided options
+  if(LLEXT_PRE_BUILD)
+    # > before the object files are linked:
+    #   - execute user command(s) before the lib target's link step.
+    set(cmd_target ${llext_lib_target})
+    set(build_step PRE_LINK)
+  elseif(LLEXT_POST_BUILD)
+    # > after linking, but before llext packaging:
+    #   - replace processor "file copy" command with a dummy to prevent
+    #     user files from being clobbered;
+    #   - execute user command(s) after the (empty) processor target.
+    set_property(TARGET ${llext_proc_target} PROPERTY ProcessorCmd
+      ${CMAKE_COMMAND} -E true
+    )
```

Comment:
not fond of substituting commands this way.

Instead of extra target creation and commands, then has it been considered if the copy here:
https://github.com/zephyrproject-rtos/zephyr/blob/21b29c3d9cc2b90b97001e2ba77caa0d196822fa/cmake/modules/extensions.cmake#L5275

simply can be depending upon the fact of whether a post build command has been defined or not.

Something like:
```
    COMMAND ${CMAKE_COMMAND} -E copy $<IF:$<BOOL:$<TARGET_PROPERTY:${llext_proc_target},${a_property_name}>>,${llext_pkg_input},${llext_lib_output}> ${llext_pkg_output}
```

In the above example, if `a_property_name` (which can be set as `LLEXT_POST_BUILD`) is defined, then the intermediate output file `llext_pkg_input` is copied to `llext_pkg_output`.

But if `a_property_name` is not defined, then `llext_lib_output` is copied directly to `llext_pkg_output`.

Thus the intermediate copy here:
https://github.com/zephyrproject-rtos/zephyr/blob/21b29c3d9cc2b90b97001e2ba77caa0d196822fa/cmake/modules/extensions.cmake#L5266 can be avoided, as well as the custom command here:
https://github.com/zephyrproject-rtos/zephyr/blob/21b29c3d9cc2b90b97001e2ba77caa0d196822fa/cmake/modules/extensions.cmake#L5258
can be removed, afaict.

Probably some more cleanup can be done, but that will depend on the exact solution.

Reply:
one more observation related to this.
When substituting the copy command with `-E true` or the similar `$<IF:$<BOOL...` construct, then it appears that it must be ensured that the custom llext build step produces `llext_pkg_input`, else later step like: `${CMAKE_COMMAND} -E copy ${llext_pkg_input} ${llext_pkg_output} ` will fail.

But I don't see that mentioned in the function description of `add_llext_command()` or being checked.
Feel free to correct me if I've missed some details in code somewhere.

Reply:
>  it must be ensured that the custom llext build step produces `llext_pkg_input` [...] I don't see that mentioned in the function description or being checked.

Missed this bit - indeed, will absolutely amend the function doc. Not sure how I could check this in code TBH :sweat_smile: